### PR TITLE
[SPARK-14052] [SQL] build a BytesToBytesMap directly in HashedRelation

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

Currently, for the key that can not fit within a long,  we build a hash map for UnsafeHashedRelation, it's converted to BytesToBytesMap after serialization and deserialization. We should build a BytesToBytesMap directly to have better memory efficiency.

In order to do that, BytesToBytesMap should support multiple (K,V) pair with the same K,  Location.putNewKey() is renamed to Location.append(), which could append multiple values for the same key (same Location). `Location.newValue()` is added to find the next value for the same key.
## How was this patch tested?

Existing tests. Added benchmark for broadcast hash join with duplicated keys.

### CODE REVIEW

**Comment by davies** [comment]
cc @JoshRosen , i'm adding more unit tests for it.

**Comment by rxin** [comment]
cc @sameeragarwal too

**Comment by JoshRosen** [comment]
Does this have any performance impact on runtime rather than peak memory usage?

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 1]
```diff
@@ -480,11 +480,29 @@ public void safeLookup(Object keyBase, long keyOffset, int keyLength, Location l
...
   /**
+   * Append a pair of key-value at the end, there could other pairs with the same keys.
+   */
+  public boolean append(Object keyBase, long keyOffset, int keyLength,
+                        Object valueBase, long valueOffset, int valueLength) {
+    assert(longArray != null);
+    int hash = Murmur3_x86_32.hashUnsafeWords(keyBase, keyOffset, keyLength, 42);
+    int pos = hash & mask;
+    int step = 1;
+    while (longArray.get(pos * 2) != 0) {
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: What guarantees that this loop will terminate?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 2]
```diff
@@ -340,6 +260,29 @@ private[joins] final class UnsafeHashedRelation(
   }
 }
 
+/**
+ * A HashedRelation for UnsafeRow with unique keys.
+ */
+private[joins] final class UniqueUnsafeHashedRelation(
+  private var numFields: Int,
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: These two lines are underindented.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 3]
```diff
@@ -555,6 +574,31 @@ private Location with(Object base, long offset, int length) {
     }
 
     /**
+     * Find the next pair that has the same key as current one.
+     */
+    public boolean next() {
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Maybe call this nextPairWithMatchingKey()?

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 4]
```diff
@@ -555,6 +574,31 @@ private Location with(Object base, long offset, int length) {
...
+      pos = (pos + step++) & mask;
+      // Remember the current key
+      Object keyBase = getKeyBase();
+      long keyOffset = getKeyOffset();
+      int keyLength = getKeyLength();
+      assert (longArray != null);
+      while (longArray.get(pos * 2) != 0) {
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: What guarantees termination here?

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java (hunk 1 above)
Comment (davies) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: It's the same as we insert new key into the map, the map can be atmost 70% full.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java (hunk 4 above)
Comment (davies) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: The same as we insert new key.

**Comment by davies** [comment]
@JoshRosen Had updated the benchmark result for BroadcastHashJoin in local mode, which use the HashMap before this patch, it become 50% slower (same for ShuffledHashJoin). The good part is that we are using less memory (also less objects), and more GC friendly, also support OffHeap, That's the same reason we use BytesToBytesMap for aggregate.

**Comment by davies** [comment]
@rxin Had switched to the algorithm you suggested, it's easier than I thought, also faster than previous one in the cases of duplicated keys (as we expected), because we do not need to compare the key for every value anymore (which is not cheap). But still have multiple copies of keys, because I want to keep the iterator as simple as before.

**Comment by davies** [comment]
@sameeragarwal PING

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 5]
```diff
@@ -560,6 +572,20 @@ private Location with(Object base, long offset, int length) {
...
     /**
+     * Find the next pair that has the same key as current one.
+     */
+    public boolean nextValue() {
+      // Remember the current key
+      assert isDefined;
+      long nextAddr = Platform.getLong(baseObject, valueOffset + valueLength);
+      if (nextAddr == 0) {
+        return false;
+      }
```
Comment (sameeragarwal) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: nit: maybe have the next 2 lines in an else block for better readability?

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 6]
```diff
@@ -657,26 +682,23 @@ public int getValueLength() {
...
      */
-    public boolean putNewKey(Object keyBase, long keyOffset, int keyLength,
-        Object valueBase, long valueOffset, int valueLength) {
-      assert (!isDefined) : "Can only set value once for a key";
-      assert (keyLength % 8 == 0);
-      assert (valueLength % 8 == 0);
-      assert(longArray != null);
-
+    public boolean append(Object kbase, long koff, int klen, Object vbase, long voff, int vlen) {
+      assert (klen % 8 == 0);
```
Comment (sameeragarwal) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: would bitwise & be faster here?

File: core/src/test/java/org/apache/spark/unsafe/map/AbstractBytesToBytesMapSuite.java [hunk 7]
```diff
@@ -587,6 +587,44 @@ public void spillInIterator() throws IOException {
   }
 
   @Test
+  public void multpleValuesForSameKey() {
```
Comment (sameeragarwal) [inline core/src/test/java/org/apache/spark/unsafe/map/AbstractBytesToBytesMapSuite.java]: nit: typo

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala [hunk 8]
```diff
@@ -184,11 +184,29 @@ class BenchmarkWholeStageCodegen extends SparkFunSuite {
...
-    Join w 2 longs codegen=false              7877 / 8358         13.3          75.1       1.0X
-    Join w 2 longs codegen=true               3877 / 3937         27.0          37.0       2.0X
+    Join w 2 longs codegen=false           12725 / 13158          8.2         121.4       1.0X
+    Join w 2 longs codegen=true              6044 / 6771         17.3          57.6       2.1X
       */
+
+    val dim4 = broadcast(sqlContext.range(M)
+      .selectExpr("cast(id/10 as long) as k1", "cast(id/10 as long) as k2"))
+
+    runBenchmark("Join w 2 longs duplicated", N) {
```
Comment (sameeragarwal) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: nit: inner join

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 9]
```diff
@@ -347,29 +293,53 @@ private[joins] object UnsafeHashedRelation {
...
+          Long.MaxValue,
+          1),
+        0)
+    }
+    val pageSizeBytes = Option(SparkEnv.get).map(_.memoryManager.pageSizeBytes)
+      .getOrElse(new SparkConf().getSizeAsBytes("spark.buffer.pageSize", "16m"))
+
+    val binaryMap = new BytesToBytesMap(
+      taskMemoryManager,
+      (sizeEstimate * 1.5 + 1).toInt, // reduce hash collision
```
Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: can you please comment here on why this particular value works best for reducing collisions?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala [hunk 10]
```diff
@@ -347,29 +293,53 @@ private[joins] object UnsafeHashedRelation {
...
+        val loc = binaryMap.lookup(key.getBaseObject, key.getBaseOffset, key.getSizeInBytes)
+        if (loc.isDefined) {
+          allUnique = false
+        }
+        val success = loc.append(
+          key.getBaseObject, key.getBaseOffset, key.getSizeInBytes,
+          row.getBaseObject, row.getBaseOffset, row.getSizeInBytes)
+        if (!success) {
+          binaryMap.free()
+          throw new SparkException("There is no enough memory to build hash map")
```
Comment (sameeragarwal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala]: nit: not

**Comment by sameeragarwal** [comment]
LGTM, just few minor comments

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java (hunk 6 above)
Comment (davies) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: assert is only used when testing, also the JIT compiler can optimize these.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala (hunk 8 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: All these are inner join, except the last two, so ignored.

**Comment by davies** [comment]
Merging this into master (since the last commit is trivial)

**Pull request closed without merging**